### PR TITLE
Rust implementation

### PR DESCRIPTION
**Pull request by kirushik**

### CODE REVIEW

**Comment by rphmeier** [comment]
You may want to specify this as Rust-Coroutine or something of that sort to avoid sending the message that this is the officially sanctioned way of performing this type of computation (not that it isn't valid!)

**Comment by kirushik** [comment]
@rphmeier Fair enough, I've updated my branch. Thanks for the tip!

**Comment by atemerev** [comment]
Will try to integrate tomorrow, among the rest of updates. :) Rust is certainly interesting to compare with Go.

File: README.md [hunk 1]
```diff
@@ -89,3 +89,11 @@ Go to `crystal/`
 ### .NET/TPL
 
 Build the solution with VS2015. Windows only :(
+
+### Rust (with [coroutine-rs](https://github.com/rustcc/coroutine-rs))
+
+```bash
+cd ./rust-coroutine
+cargo build --release
```
Comment (crumblingstatue) [inline README.md:97]: `cargo run --release` alone is just fine. It will build the target if it's out of date.

Comment (kirushik) [inline README.md:97]: Thanks!
I just wanted to make that explicit so noone would include compilation time in his benchmarks.

Comment (crumblingstatue) [inline README.md:97]: > I just wanted to make that explicit so noone would include compilation time in his benchmarks.

Oh, I see, makes sense.

**Comment by atemerev** [comment]
Please advise:

~/d/g/rust-coroutine ⇢  cargo build --release
    Updating git repository `https://github.com/rustcc/coroutine-rs.git`
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `http://github.com/zonyitoo/context-rs.git`
 Downloading libc v0.1.12
 Downloading memmap v0.2.3
 Downloading kernel32-sys v0.2.1
 Downloading log v0.3.5
 Downloading winapi-build v0.1.1
 Downloading libc v0.2.7
 Downloading winapi v0.2.5
 Downloading gcc v0.3.24
 Downloading fs2 v0.2.2
   Compiling winapi-build v0.1.1
   Compiling winapi v0.2.5
   Compiling gcc v0.3.24
   Compiling libc v0.2.7
   Compiling kernel32-sys v0.2.1
   Compiling libc v0.1.12
/Users/atemerev/.cargo/registry/src/github.com-88ac128001ac3a9a/libc-0.1.12/rust/src/liblibc/lib.rs:81:21: 81:39 warning: lint raw_pointer_derive has been removed: using derive with raw pointers is ok
/Users/atemerev/.cargo/registry/src/github.com-88ac128001ac3a9a/libc-0.1.12/rust/src/liblibc/lib.rs:81 #![allow(bad_style, raw_pointer_derive)]
                                                                                                                           ^~~~~~~~~~~~~~~~~~
   Compiling fs2 v0.2.2
   Compiling log v0.3.5
   Compiling memmap v0.2.3
   Compiling context v0.2.1 (http://github.com/zonyitoo/context-rs.git#2c3c33cc)
/Users/atemerev/.cargo/git/checkouts/context-rs-44ad4be5ac90aec5/master/src/lib.rs:1:1: 1:28 error: #[feature] may not be used on the stable release channel
/Users/atemerev/.cargo/git/checkouts/context-rs-44ad4be5ac90aec5/master/src/lib.rs:1 #![feature(asm, repr_simd)]
                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
Could not compile `context`.

**Comment by kirushik** [comment]
@atemerev Oh, looks like we should state that in the readme — it will only work with the nightly Rust. I'm sorry for not noticing that in advance.

Rust is a relatively new language, and many of it's features are not stable yet.
Those are hidden behind feature-flags, like the `#![feature(asm, repr_simd)]` on in the [context-rs](github.com/zonytoo/context-rs) library, which my code indirectly uses.
But stable versions of compiler are all about backwards-compatibility, and are unable to compile the code with feature-flagged features. (Even some basic parts of standard IO library were incomplete and feature-flagged in Rust 1.0 stable!)

It's trivial to install the nightly version of Rust, as seen on https://www.rust-lang.org/downloads.html
You just need to run

``` rust
curl -sSf https://static.rust-lang.org/rustup.sh | sh -s -- --channel=nightly
```

Or use something like [rsvm](github.com/sdepold/rsvm) to have several versions of Rust simpultaneously installed on the same machine.

File: README.md (hunk 1 above)
Comment (Veedrac) [inline README.md:97]: `time cargo run --release` takes noticeably longer (~10-20%) than `time ./target/release/rust`.

File: rust-coroutine/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+fn skynet(me: coroutine::asymmetric::CoroutineRef<(u64, u64)>) {
+    let (my_number, remaining) = me.take_data().unwrap();
+
+    if remaining==1 {
+        me.yield_with((my_number, 0));
+    } else {
+        let mut coros: Vec<coroutine::asymmetric::Coroutine<(u64, u64)>> = Vec::with_capacity(10);
+
+        for _ in 0..10 {
+            coros.push(Coroutine::spawn(&skynet));
```
Comment (Veedrac) [inline rust-coroutine/src/main.rs:14]: More idiomatic yet practically equivalent is

```
let coros: Vec<_> = (0..10).map(|_| Coroutine::spawn(&skynet)).collect();
```

**Pull request merged**